### PR TITLE
[VersionControl] Synchronize VersionControlService cache access

### PR DESCRIPTION
**Pull request by sevoku**
Fixes `Could not query Git repository reference` errors and avoids creating several repository instances in parallel, where only one is required to be reused from the cache later.

Fixes VSTS #806706

### CODE REVIEW

**Comment by sevoku** [comment]
@monojenkins backport release-8.0-integration

**Comment by slluis** [comment]
@sevoku can you fix the conflicts?

**Comment by slluis** [comment]
It looks like it has changes from other PRs. Also, is there any specific bug that this fixes?

**Comment by sevoku** [comment]
oh, looks like I messed up the commit somehow, bit GH showed the right changes initially... let me recommit

**Comment by sevoku** [comment]
@monojenkins backport release-8.0-integration

File: main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/VersionControlService.cs [hunk 1]
```diff
@@ -249,18 +247,21 @@ public static Repository GetRepositoryReference (string path, string id)
...
+						return result;
+					}
+					// never add null values
+					throw new ArgumentNullException ("result");
+				});
 			} catch (Exception e) {
-				LoggingService.LogError ($"Could not query {detectedVCS.Name} repository reference", e);
+				// ArgumentNullException for "result" is expected when GetRepositoryReference returns null, no need to log
+				if (!(e is ArgumentNullException ne) || ne.ParamName != "result")
+					LoggingService.LogError ($"Could not query {detectedVCS.Name} repository reference", e);
```
Comment (Therzok) [inline main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/VersionControlService.cs]:
```suggestion
					LoggingService.LogInternalError ($"Could not query {detectedVCS.Name} repository reference", e);
```

Comment (slluis) [reply main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/VersionControlService.cs]: @sevoku can you apply this change?

Comment (Therzok) [reply main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/VersionControlService.cs]: @slluis Fixed it, if you temporarily assign yourself, you can commit the suggestion

**Comment by slluis** [comment]
@monojenkins backport release-8.0-integration

**Comment by Therzok** [comment]
@monojenkins backport release-8.0-integration

**Pull request merged**